### ISSUE TITLE
Service container got recreated after "docker service update --image" with image has same repo digest but different tag

### ISSUE THREAD

**Issue by intenseFrog** [moby/moby#33905]
**Description**
Service container got recreated when using _"docker service update"_ even though the image used for _"docker service update"_ is the **same image** used during _"docker service update --image"_. By **same image**, I mean the two images shared **identical repo digest** but under **different tag**. For example: image nginx:latest and nginx:1.13 shared identical repo digest but with different tags.

I wonder if it is a bug or by design? Is it possible to add a flag like "--no-resolve-image" but to only resolve the repo digest?

**Steps to reproduce the issue:**
1. docker service create --name nginx nginx:latest
2. docker service update --image nginx:1.13 nginx
3. docker service ps nginx
3. docker service inspect nginx

**Describe the results you received:**
the nginx container has been recreated due to "docker service update"

**Describe the results you expected:**
the nginx container shouldn't have been recreated since "nginx:latest" and "nginx:1.13" are two same images(same digest) only with different tags. 

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:15:15 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 17.06.0-ce
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 7
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: hkeop8rz0yy1rvv7wdim0bg3q
 Is Manager: true
 ClusterID: pucz4cvrtbabkm1tckli698e8
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.99.108
 Manager Addresses:
  192.168.99.108:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.74-boot2docker
Operating System: Boot2Docker 17.06.0-ce (TCL 7.2); HEAD : 0672754 - Thu Jun 29 00:06:31 UTC 2017
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.8MiB
Name: 06
ID: 6RBH:UMRI:QQ6P:J53X:TOSJ:QZUT:6N4K:QCTA:ZV5Y:W2GE:SRAZ:HQMZ
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 39
 Goroutines: 151
 System Time: 2017-06-30T21:23:16.426357468Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Experimental: false
Insecure Registries:
 10.10.1.181:5000
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
Thanks for reporting and for providing a clear description and reproduction steps.

I reproduced your issue, and it looks to be a duplicate of https://github.com/moby/moby/issues/33865, which has been fixed by https://github.com/moby/moby/pull/33867, and will be included in a patch release of docker (docker 17.06.1), that is already being prepared.

These are the steps I used to reproduce;

```bash
$ docker pull nginx:latest
...
Digest: sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702
Status: Downloaded newer image for nginx:latest

$ docker pull nginx:1.13
...
Digest: sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702
Status: Downloaded newer image for nginx:1.13


$ docker service create --name nginx nginx:latest
$ docker service inspect nginx > before.json

$ docker service update --image nginx:1.13 nginx
$ docker service inspect nginx > after.json
```

Comparing the before and after

```bash
$ git diff before.json after.json
```

Produced the following diff, which confirms this is the same issue as https://github.com/moby/moby/issues/33865

```patch
diff --git a/before.json b/after.json
index f931286..310729e 100644
--- a/before.json
+++ b/after.json
@@ -2,16 +2,16 @@
     {
         "ID": "w71fcsu5ubo7g0wt5ahd4y90a",
         "Version": {
-            "Index": 10106
+            "Index": 10124
         },
         "CreatedAt": "2017-07-02T04:07:00.746248069Z",
-        "UpdatedAt": "2017-07-02T04:07:00.746248069Z",
+        "UpdatedAt": "2017-07-02T04:07:51.446484178Z",
         "Spec": {
             "Name": "nginx",
             "Labels": {},
             "TaskTemplate": {
                 "ContainerSpec": {
-                    "Image": "nginx:latest@sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702",
+                    "Image": "nginx:1.13@sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702",
                     "StopGracePeriod": 10000000000,
                     "DNSConfig": {}
                 },
@@ -29,6 +29,10 @@
                         {
                             "Architecture": "amd64",
                             "OS": "linux"
+                        },
+                        {
+                            "Architecture": "amd64",
+                            "OS": "linux"
                         }
                     ]
                 },
@@ -58,8 +62,46 @@
                 "Mode": "vip"
             }
         },
+        "PreviousSpec": {
+            "Name": "nginx",
+            "Labels": {},
+            "TaskTemplate": {
+                "ContainerSpec": {
+                    "Image": "nginx:latest@sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702",
+                    "DNSConfig": {}
+                },
+                "Resources": {
+                    "Limits": {},
+                    "Reservations": {}
+                },
+                "Placement": {
+                    "Platforms": [
+                        {
+                            "Architecture": "amd64",
+                            "OS": "linux"
+                        }
+                    ]
+                },
+                "ForceUpdate": 0,
+                "Runtime": "container"
+            },
+            "Mode": {
+                "Replicated": {
+                    "Replicas": 1
+                }
+            },
+            "EndpointSpec": {
+                "Mode": "vip"
+            }
+        },
         "Endpoint": {
             "Spec": {}
+        },
+        "UpdateStatus": {
+            "State": "completed",
+            "StartedAt": "2017-07-02T04:07:43.453737843Z",
+            "CompletedAt": "2017-07-02T04:07:51.446463335Z",
+            "Message": "update completed"
         }
     }
 ]
```

I'm closing this issue, because this has been fixed on master, but thanks again for reporting, and feel free to continue the discussion :+1:

**Comment by thaJeztah** [comment]
hm, actually, it will probably still re-deploy, because the actual image information changes;

```patch
-                    "Image": "nginx:latest@sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702",
+                    "Image": "nginx:1.13@sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702",
```

I think that is by design (i.e., the service spec changed), but perhaps there's an option to ignore the tag if the sha matches

Let me reopen for discussion

ping @aaronlehmann @nishanttotla

**Comment by nishanttotla** [comment]
The design decision is that if the digest came from the user (or client) then then it should be treated as immutable regardless of tag. But when you update the image tag, the spec is technically updated and the service will update as well.

It might be worthwhile to not update the service if the digest hasn't changed (this is easy to do). Since `docker pull` on the agents will pull the same image as long as the digest hasn't changed, this keeps the tasks uniform. I'm not sure that this is a particularly common use case though, so we should confirm that.

**Comment by thaJeztah** [comment]
> It might be worthwhile to not update the service if the digest hasn't changed (this is easy to do)

Yes, I was thinking along those lines, but have to admit it may be a bit of a corner-case

Let's see what @aaronlehmann thinks 👍

**Comment by aaronlehmann** [comment]
> It might be worthwhile to not update the service if the digest hasn't changed (this is easy to do)

I don't think we should make this change. If the user asks for the tag to be updated, I think the right thing to do is update the tag. It would be strange to ignore this request because the client knows the tags map to the same digest.

I'm not sure what the problem is here. If a user asks to switch to a different image, it seems reasonable to recreate the containers.

**Comment by thaJeztah** [comment]
@aaronlehmann is it possible to update the spec without re-reploying tasks? (i.e., swarm understanding that `image:foo@sha256:deadbeef` and `image:bar@sha256:deadbeef` are the same image, because they have the same sha)

**Comment by intenseFrog** [comment]
@nishanttotla @aaronlehmann @nishanttotla 
The scenario is following: we use git commit ID as image tag. Each commit of the git project will trigger a build of **multiple** images. Therefore, each image under the project will have a new tag after a commit regardless of its change. 

The problem is that: our CI/CD process will then deploy/update the stack using new image tags. 
We don't want those containers that are using same images to be re-deployed though with new tags. 

So I'm wondering if there could be a flag to solve this issue. Or what would you recommend to be a good approach of solving this? Thanks

**Comment by aaronlehmann** [comment]
> is it possible to update the spec without re-reploying tasks? (i.e., swarm understanding that `image:foo@sha256:deadbeef` and `image:bar@sha256:deadbeef` are the same image, because they have the same sha)

Technically we could do that. I'm hesitant to introduce exceptions. It's very simple to understand that right now if the `ContainerSpec`s are the same, the service is not redeployed, otherwise it is. When we've tried in the past to support things like order-insensitivity of certain fields, things got really complicated, and we ended up not going down this path.

Also, we're trying to remove all logic around `ContainerSpec` in swarmkit. Ideally this should just be an opaque set of fields that get passed through from the API client to the executor. In the future we might even move the definition outside of swarmkit.

> So I'm wondering if there could be a flag to solve this issue. Or what would you recommend to be a good approach of solving this?

It would definitely be doable as a CLI flag, but the flag might be weird from a UX perspective.

Would it be possible for your deployment script to check if the sha256 hash matches the previous value, and avoid updating the service in this case? If there is a `docker push` step, that should output the sha256 hash.

**Comment by intenseFrog** [comment]
I understand the complexity of it. But I would still argue that users should be allowed to decide whether they want to re-deploy services based on image tag or sha256 hash.

> Would it be possible for your deployment script to check if the sha256 hash matches the previous value, and avoid updating the service in this case? 

Yes, it will be doable. Right now we are using `docker stack deploy --compose-file` to deploy/update services, will then need to check sha256 hash and update compose file accordingly.

**Comment by korigod** [comment]
> Right now we are using `docker stack deploy --compose-file` to deploy/update services

It's really nice way to update the services — much more convenient and straight-forward than additional script to check if every service must be updated and issue `docker service update` commands. Now the only nice way is to use a generic tag like `latest` or `dev` but it's less error-prone.

Is there a probability that the discussed option will be implemented for `stack deploy` and `service update` commands?
